File: exercise/ts/day07/src/workshop/workshop.ts
import {Gift} from "./gift";
import {Status} from "./status";

export class Workshop {
    private gifts: Gift[] = [];

    addGift(gift: Gift): void {
        this.gifts.push(gift);
    }

    completeGift(name: string): Gift | null {
        const gift = this.gifts.find(gift => gift.getName() === name);
        return gift ? gift.withStatus(Status.Produced) : null;
    }
}

File: exercise/java/day07/src/main/java/workshop/Workshop.java
package workshop;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Workshop {
    private final List<Gift> gifts = new ArrayList<>();

    public void addGift(Gift gift) {
        gifts.add(gift);
    }

    public Optional<Gift> completeGift(String name) {
        return gifts.stream()
                .filter(gift -> gift.name().equals(name))
                .findFirst()
                .map(gift -> gift.withStatus(Status.PRODUCED));
    }
}
